### PR TITLE
Replace voluptuous with pydantic

### PR DESCRIPTION
**Pull request by dbluhm**

### CODE REVIEW

File: pydid/doc/builder.py [hunk 1]
```diff
@@ -102,19 +83,6 @@ def _default_id_generator(self):
         )
         yield from _default_id_generator(self._id_base, start=start)
```
Comment (burdettadam) [inline pydid/doc/builder.py:84]: should this be `self._id_generator(`? Using two different id generators will create unintentionally duplicate ids.

File: pydid/doc/corrections.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        nonlocal index
+        if isinstance(value, str):
+            return
+        if isinstance(value, list):
+            for item in value:
+                _walk(item)
+            return
+        if isinstance(value, dict):
+            if "id" not in value:
+                value["id"] = str(did.ref("inserted-{}".format(index)))
```
Comment (burdettadam) [inline pydid/doc/corrections.py:41]: consider using an f-string, like `f'inserted-{index}'`.

File: tests/doc/test_doc.py [hunk 3]
```diff
@@ -66,162 +81,162 @@
...
+            "type": "Ed25519Verification2018",
             "controller": "did:example:z6Mkmpe2DyE4NsDiAb58d75hpi1BjqbH6wYMschUkjWDEEuR",
             "publicKeyBase58": "8NNydiyd3KjF46ERwY7rycTBvGKRh4J1BbnYvTYCK283",
         }
     ],
     "capabilityInvocation": [
         {
             "id": "did:example:123#capability-invocation-1",
-            "type": "Ed25519VerificationKey2018",
+            "type": "Ed25519Verification2018",
```
Comment (burdettadam) [inline tests/doc/test_doc.py:173]: This changes the interface of the tests being called. Be careful.

**Review by burdettadam — APPROVED** [review]

File: pydid/did.py [hunk 4]
```diff
@@ -41,34 +54,13 @@ def method_specific_id(self):
...
-
     def url(self, path: str = None, query: Dict[str, str] = None, fragment: str = None):
         """Return a DID URL for this DID."""
-        return DIDUrl(self._raw, path, query, fragment)
+        return DIDUrl.unparse(self, path, query, fragment)
 
     def ref(self, ident: str) -> DIDUrl:
         """Return a DID reference (URL) for use as an ID in a DID Doc section."""
-        return DIDUrl(self._raw, fragment=ident)
+        return DIDUrl.unparse(self, fragment=ident)
```
[the remark this answers is not in the document]
Comment (dbluhm) [reply pydid/did.py:63]: Good point; my reasoning was that the parse and unparse mirror methods of the same names in pythons `urllib.parse` module. With that context, does unparse feel less weird? lol

**Review by burdettadam — APPROVED** [review]

File: pydid/doc/builder.py (hunk 1 above)
Comment (dbluhm) [reply pydid/doc/builder.py:84]: This is saying "yield from the global _default_id_generator initialized with these values" so delegating to a generator instance that is created in that method

Comment (dbluhm) [reply pydid/doc/builder.py:84]: i.e. `_default_id_generator` from line 13 vs `self._default_id_generator`

**Pull request merged**